delta_shooter/models.py

```diff
@@ -0,0 +1,277 @@
+        self.position = new_position
+
+    def draw(self, surface: pygame.surface.Surface) -> None:
+        angle = self.direction.angle_to(UP)
+        assert isinstance(self.sprite, pygame.Surface)
+        rotated_surface = rotozoom(self.sprite, angle, 1.0)
+        rotated_surface_size = Vector2(rotated_surface.get_size())
+        blit_position = self.position - rotated_surface_size * 0.5
+        surface.blit(rotated_surface, blit_position)
+
+    def shoot(self) -> None:
+        # Limit number of bullets
+        if len(self.bullets) == self.NUM_BULLETS:
+            return
+        bullet_velocity = self.direction * self.BULLET_SPEED + self.velocity
+        bullet = Bullet(self.position, bullet_velocity, self.graphical)
+        self.bullets.append(bullet)
+        self.laser_sound.play()
+
+
+class Bullet(GameObject):
+    def __init__(
+        self,
+        position: Union[Tuple[int, int], Vector2],
+        velocity: Union[int, Vector2],
+        graphical: bool,
+    ):
+        self.hit_barrier = False
+        if graphical:
+            super().__init__(position, load_sprite("bullet"), velocity)
+        else:
+            super().__init__(position, DummyBullet(), velocity)
+        self.name = "bullet"
+
+    def move(self, surface: Any) -> None:
+        new_position = self.position + self.velocity
+        for barrier in BARRIERS:
+            if barrier.hit_barrier(self.position, new_position, self.radius):
+                if barrier.orientation == "vertical":
+                    self.set_position((barrier.center[0], int(new_position[1])))
+                    self.hit_barrier = True
+                    return
+                elif barrier.orientation == "horizontal":
+                    self.set_position((int(new_position[0]), barrier.center[1]))
+                    self.hit_barrier = True
+                    return
+        self.set_position(new_position)
+
+
+class Barrier:
+    def __init__(self, orientation: Orientation, length: int, center: Tuple[int, int]):
+        self.orientation = orientation
+        # Prevents mypy isssues with potentially unbound vars
+        assert orientation in {"vertical", "horizontal"}, "Invalid Orientation"
+        self.length = length
+        self.center = center
+        self.name = "barrier"
+
+        if orientation == "vertical":
+            self.end1 = (self.center[0], self.center[1] - self.length // 2)
+            self.end2 = (self.center[0], self.center[1] + self.length // 2)
+        else:
+            self.end1 = (self.center[0] - self.length // 2, self.center[1])
+            self.end2 = (self.center[0] + self.length // 2, self.center[1])
+
+    def hit_barrier(
+        self,
+        pos: Union[Tuple[int, int], Vector2],
+        new_pos: Union[Tuple[int, int], Vector2],
+        radius: int,
+    ) -> bool:
+        """There's probably a way to generalise this to diagonal barriers with linear algebra but
+        cba.
+
+        (wow this function is janky)
```

Comment:
Delete this line and add a ticket to backlog to fix this?

Reply:
Should be covered by this

https://learney.atlassian.net/browse/LRN-856?atlOrigin=eyJpIjoiNjYwMThlZDI3NjVkNGE3OGJiNWQzNjUzNWNjNThmNTkiLCJwIjoiaiJ9